testutils/src/main/java/com/microsoft/identity/internal/testutils/CacheUtils.java

```diff
@@ -0,0 +1,147 @@
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.internal.testutils;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.text.TextUtils;
+
+import androidx.annotation.NonNull;
+import androidx.arch.core.util.Function;
+import androidx.test.core.app.ApplicationProvider;
+
+import com.microsoft.identity.common.internal.cache.CacheKeyValueDelegate;
+import com.microsoft.identity.common.internal.cache.SharedPreferencesAccountCredentialCache;
+import com.microsoft.identity.common.internal.dto.Credential;
+import com.microsoft.identity.common.internal.dto.CredentialType;
+
+import java.util.Map;
+import java.util.Random;
+
+public class CacheUtils {
+
+    private static final CacheKeyValueDelegate CACHE_KEY_VALUE_DELEGATE = new CacheKeyValueDelegate();
+
+    public static final Function<String, String> TOKEN_SIGNATURE_EDITOR = new Function<String, String>() {
+        @Override
+        public String apply(String s) {
+            return editTokenSignature(s);
+        }
+    };
+
+    private interface Predicate<T> {
+        boolean test(T t);
+    }
+
+    /**
+     * This method will edit all the token specified by the predicate using the  editor
+     * in the shared preference.
+     *
+     * @param sharedPrefName Name of the shared preference where token has been stored.
+     * @param predicate      Generic functional interface representing function that returns true
+     *                       or false depending on taken type.
+     * @param editor         Functional interface to have any number of token editing method.
+     */
+    public void editAllTokenInCache(@NonNull final String sharedPrefName, Predicate<String> predicate, Function<String, String> editor) {
+        SharedPreferences sharedPref = TestUtils.getSharedPreferences(sharedPrefName);
+        SharedPreferences.Editor prefEditor = sharedPref.edit();
+        Map<String, ?> cacheEntries = sharedPref.getAll();
+
+        //get all the key from the cache entry, verify and edit it.
+        for (Map.Entry<String, ?> cacheEntry : cacheEntries.entrySet()) {
+            String keyToEdit = cacheEntry.getKey();
+            if (predicate.test(keyToEdit)) {
+                Object cacheValue = cacheEntries.get(keyToEdit);
+                if (cacheValue instanceof String) {
```

Comment:
cache value is string when stored in the shared preference therefore we would get string whenever we fetch cache value form the cache.

Reply:
why we are checking it then?

Reply:
sharedPeference getall() method returns a Map<String, ?> containing list of key/value pairs representing the preference.
"?" which represent any object therefore the Map maps string key to any possible object value. Here our object is of string because both key/value are string when stored in the shared preference. Moreover check has been added to ensure only tokens(value are of string object) should be fetched.

Reply:
We should probably just let it throw a ClassCastException rather than trying to be clever.  This is a test utility.

Reply:
Make sense. cast cacheValue to string.